fission/src/systems/physics/PhysicsSystem.ts

```diff
@@ -1005,6 +1022,96 @@ class PhysicsSystem extends WorldSystem {
             activate ? JOLT.EActivation_Activate : JOLT.EActivation_DontActivate
         )
     }
+
+    /**
+     * Exposes SetShape method on the _joltBodyInterface
+     * Sets the shape of the body
+     *
+     * @param id The id of the body
+     * @param shape The new shape of the body
+     * @param massProperties The mass properties of the new body
+     * @param activationMode The activation mode of the new body
+     */
+    public SetShape(
+        id: Jolt.BodyID,
+        shape: Jolt.Shape,
+        massProperties: boolean,
+        activationMode: Jolt.EActivation
+    ): void {
+        if (!this.IsBodyAdded(id)) {
+            return
+        }
+
+        this._joltBodyInterface.SetShape(id, shape, massProperties, activationMode)
+    }
+
+    /**
+     * Creates and assigns Jolt contact listener that dispatches events.
+     *
+     * @param physSystem The physics system the contact listener will attach to
+     */
+    private SetUpContactListener(physSystem: Jolt.PhysicsSystem) {
+        const contactListener = new JOLT.ContactListenerJS()
+
+        contactListener.OnContactAdded = (bodyPtr1, bodyPtr2, manifoldPtr, settingsPtr) => {
+            const body1 = JOLT.wrapPointer(bodyPtr1, JOLT.Body) as Jolt.Body
+            const body2 = JOLT.wrapPointer(bodyPtr2, JOLT.Body) as Jolt.Body
+
+            const body1Id = new JOLT.BodyID(body1.GetID().GetIndexAndSequenceNumber())
+            const body2Id = new JOLT.BodyID(body2.GetID().GetIndexAndSequenceNumber())
+
+            const message: CurrentContactData = {
+                body1: body1Id,
+                body2: body2Id,
+                manifold: JOLT.wrapPointer(manifoldPtr, JOLT.ContactManifold) as Jolt.ContactManifold,
+                settings: JOLT.wrapPointer(settingsPtr, JOLT.ContactSettings) as Jolt.ContactSettings,
+            }
+
+            this._physicsEventQueue.push(new OnContactAddedEvent(message))
+        }
+
+        contactListener.OnContactPersisted = (bodyPtr1, bodyPtr2, manifoldPtr, settingsPtr) => {
+            const body1 = JOLT.wrapPointer(bodyPtr1, JOLT.Body) as Jolt.Body
+            const body2 = JOLT.wrapPointer(bodyPtr2, JOLT.Body) as Jolt.Body
+
+            const body1Id = body1.GetID()
+            const body2Id = body2.GetID()
+
+            const message: CurrentContactData = {
+                body1: body1Id,
+                body2: body2Id,
+                manifold: JOLT.wrapPointer(manifoldPtr, JOLT.ContactManifold) as Jolt.ContactManifold,
+                settings: JOLT.wrapPointer(settingsPtr, JOLT.ContactSettings) as Jolt.ContactSettings,
+            }
+
+            this._physicsEventQueue.push(new OnContactPersistedEvent(message))
+        }
+
+        contactListener.OnContactRemoved = subShapePairPtr => {
+            const shapePair = JOLT.wrapPointer(subShapePairPtr, JOLT.SubShapeIDPair) as Jolt.SubShapeIDPair
+
+            const event = new OnContactRemovedEvent(shapePair)
+            event.Dispatch()
```

Comment:
Could you put event.dispatch inside the constructor of OnContactRemovedEvent since its not in the event queue

Reply:
@a-crowell Was having this event be dispatched in the queue causing an issue? If not, it should probably be added to the queue as well.

Reply:
Yeah having it in the queue causes issues. It seems there is just enough delay to allow removed events to go unnoticed/in the wrong order. I can put the dispatch inside the constructor though.